relayer/chain/substrate/chain.go

```diff
@@ -14,14 +14,16 @@ import (
 	"github.com/snowfork/polkadot-ethereum/relayer/chain"
 	"github.com/snowfork/polkadot-ethereum/relayer/chain/ethereum"
 	"github.com/snowfork/polkadot-ethereum/relayer/crypto/sr25519"
+	"github.com/snowfork/polkadot-ethereum/relayer/store"
 )
 
 type Chain struct {
-	config   *Config
-	listener *Listener
-	writer   *Writer
-	conn     *Connection
-	log      *logrus.Entry
+	config    *Config
+	listener  *Listener
+	writer    *Writer
+	conn      *Connection
+	relayconn *Connection
```

Comment:
In a previous version of this work, I vaguely recall you had the relay chain stuff in a separate chain abstraction called `relayer/chain/parachain`. While not correctly named, it was still the right approach. Was there a reason you moved away from that?

I don't think we should go down this path of making the substrate chain module handle both the parachain and relaychain. This wasn't really how I designed the chain abstractions to work, and adds extra complexity to a module which is already quite complicated and brittle as it is. 

Instead I suggest the following module layout for the chains involved.
* relayer/chain/ethereum
* relayer/chain/parachain
* relayer/chain/relaychain
 
There will some duplicated code, but we can improve that post-launch.

Reply:
Originally the third chain abstraction was primarily used as a dev environment of sorts as I wanted to get everything working before prematurely refactoring the existing code. It hosted all the new ethereum functionality as well, which needed to be refactored into the ethereum reader/writer. As part of that refactor I also moved the substrate-compatible portions into the substrate chain abstraction.

I agree with your suggestion, each chain should have its own independent abstraction. I'll make the suggested changes. It's worth noting that the `relayer/chain/relaychain` abstraction wouldn't need a writer, as we only read information from the relay chain via subscriptions and queries.

Reply:
yeh i was imagining the substrate abstraction to be more like a library that is used by separate relaychain/parachain abstractions rather than a module that handles both together

Reply:
I went with @vgeddes's suggestion in https://github.com/Snowfork/polkadot-ethereum/pull/314/commits/b428332f96d28dd075dd6dee02c99b9f1b554f5d, https://github.com/Snowfork/polkadot-ethereum/pull/314/commits/9251d5b526bf056a4094d3f29f06e9dec2d0bfa6. The relaychain actually ended up being quite small once all the parachain-specific logic was removed. The `Chain` struct for the two chains is similar (only difference is that relaychain doesn't have a writer) so it seems sensible to base both abstractions off a shared Substrate chain struct. I'll come back to @musnit's suggestion once some of the other revisions are addressed.

Reply:
Great, thanks!